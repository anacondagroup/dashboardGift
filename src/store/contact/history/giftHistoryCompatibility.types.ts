/* eslint-disable camelcase */

export type THistoryItem = {
  action_date: string;
  label: string;
  status_string: string;
};

export type TMeetingInfo = {
  id: number;
  additional_info: string;
  end: string;
  meeting_type: string;
  outcome: string;
  start: string;
  status: string;
  title: string;
};

export type TSender = {
  id: number;
  email: string;
  full_name: string;
  avatar: string;
};

export type TShippingInfo = {
  tracking_link: string;
  tracking_number: string;
  tracking_state: string;
  tracking_carrier: string;
};

export type TGiftHistoryItem = {
  id: number;
  campaign: {
    id: number;
    name: string;
    team_id: number;
  };
  history: THistoryItem[];
  meeting: TMeetingInfo;
  product: {
    name: string;
    image: string;
  };
  sentAs: TSender;
  sendBy: TSender;
  shippingInfo: TShippingInfo;
  statusId: number;
  createdAt: string;
  meetingNotes: string;
};
/* eslint-enable camelcase */
