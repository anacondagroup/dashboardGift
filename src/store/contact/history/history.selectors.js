import * as R from 'ramda';
import { updateSearch } from '@alycecom/modules';
import { pickNotEmpty, renameKeys } from '@alycecom/utils';

import { getContactState } from '../contact.selectors';
import { viewLoading } from '../../../helpers/lens.helpers';

export const getHistoryState = R.compose(R.prop('history'), getContactState);

export const getGifts = R.compose(R.prop('gifts'), getHistoryState);

export const getHistoryIsLoading = R.compose(viewLoading, getHistoryState);

const giftApiParams = R.pick(['teamId', 'campaignId', 'dateRangeFrom', 'dateRangeTo', 'memberId']);

export const getApiCallQuery = R.compose(
  R.curryN(2, updateSearch)(''),
  renameKeys({
    memberId: 'team_member_id',
    teamId: 'team_id',
    campaignId: 'campaign_id',
    dateRangeFrom: 'date_range_from',
    dateRangeTo: 'date_range_to',
  }),
  pickNotEmpty,
  giftApiParams,
);

export const buildApiUrl = ({ contactId, ...rest }) =>
  `/enterprise/contact/${contactId}/gift-history?${getApiCallQuery(rest)}`;
