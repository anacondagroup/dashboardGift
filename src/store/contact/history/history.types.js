export const HISTORY_LOAD_REQUEST = 'HISTORY_LOAD_REQUEST';
export const HISTORY_LOAD_SUCCESS = 'HISTORY_LOAD_SUCCESS';
export const HISTORY_LOAD_FAIL = 'HISTORY_LOAD_FAIL';
