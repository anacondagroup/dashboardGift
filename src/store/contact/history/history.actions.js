import { HISTORY_LOAD_REQUEST, HISTORY_LOAD_SUCCESS, HISTORY_LOAD_FAIL } from './history.types';

export const historyLoadRequest = contactId => ({
  type: HISTORY_LOAD_REQUEST,
  payload: contactId,
});

export const historyLoadSuccess = history => ({
  type: HISTORY_LOAD_SUCCESS,
  payload: history,
});

export const historyLoadFail = error => ({
  type: HISTORY_LOAD_FAIL,
  payload: error,
});
