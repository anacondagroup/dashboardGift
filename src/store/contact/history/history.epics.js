import { ofType } from 'redux-observable';
import { catchError, map, mergeMap } from 'rxjs/operators/index';
import { of } from 'rxjs';

import { HISTORY_LOAD_REQUEST } from './history.types';
import { historyLoadSuccess, historyLoadFail } from './history.actions';
import { buildApiUrl } from './history.selectors';

export const loadHistoryEpic = (action$, state$, { apiService }) =>
  action$.pipe(
    ofType(HISTORY_LOAD_REQUEST),
    mergeMap(action =>
      apiService.get(buildApiUrl(action.payload)).pipe(
        map(response => historyLoadSuccess(response.gifts)),
        catchError(error => of(historyLoadFail(error))),
      ),
    ),
  );

export const historyEpics = [loadHistoryEpic];
