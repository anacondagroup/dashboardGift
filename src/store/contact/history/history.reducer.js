import * as R from 'ramda';
import { renameKeys } from '@alycecom/utils';

import { setLoading } from '../../../helpers/lens.helpers';

import { HISTORY_LOAD_REQUEST, HISTORY_LOAD_SUCCESS, HISTORY_LOAD_FAIL } from './history.types';

export const initialState = {
  isLoading: false,
  error: null,
  gifts: [],
};

export const mapGift = R.compose(
  renameKeys({
    sent_as: 'sentAs',
    sent_by: 'sentBy',
    shipping_info: 'shippingInfo',
    status_id: 'statusId',
    meeting_notes: 'meetingNotes',
  }),
);

export const reducer = (state = initialState, action) => {
  switch (action.type) {
    case HISTORY_LOAD_REQUEST:
      return {
        ...state,
        ...setLoading(true, state),
      };
    case HISTORY_LOAD_SUCCESS:
      return {
        ...state,
        ...setLoading(false, state),
        gifts: R.map(mapGift, action.payload),
        error: null,
      };
    case HISTORY_LOAD_FAIL:
      return {
        ...state,
        ...setLoading(false, state),
        error: action.payload,
      };
    default:
      return state;
  }
};
