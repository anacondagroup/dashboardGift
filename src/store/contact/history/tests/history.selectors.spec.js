import { buildApiUrl } from '../history.selectors';

describe('history.selector', () => {
  describe('buildApiUrl', () => {
    it('Should build url', () => {
      const actual = buildApiUrl({
        memberId: '1',
        teamId: '1',
        campaignId: '1',
        dateRangeFrom: '11/11/2019',
        dateRangeTo: '12/12/2019',
      });

      expect(actual).toMatchSnapshot();
    });
  });
});
