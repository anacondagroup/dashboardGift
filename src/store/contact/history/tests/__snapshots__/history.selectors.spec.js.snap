// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`history.selector buildApiUrl Should build url 1`] = `"/enterprise/contact/undefined/gift-history?campaign_id=1&date_range_from=11%2F11%2F2019&date_range_to=12%2F12%2F2019&team_id=1&team_member_id=1"`;
